src/main/java/org/candlepin/subscriptions/product/MidProduct.java

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.product;
+
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.candlepin.subscriptions.db.model.Offering;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.Usage;
+import org.candlepin.subscriptions.product.api.model.AttributeValue;
+import org.candlepin.subscriptions.product.api.model.EngineeringProduct;
+import org.candlepin.subscriptions.product.api.model.OperationalProduct;
+import org.candlepin.subscriptions.product.api.model.RESTProductTree;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides an easier way to translate an operational product, its children, and their derived
+ * products into an Offering. The upstream product data are put into this intermediate structure
+ * which can then be "merged down" into its ultimate Offering form.
+ */
+public class MidProduct {
```

Comment:
Rather than converting directly from upstream's representation of products into swatch's Offering type, I introduced a representation that sits between the two. It stores all of the fields and attributes in a Map so that it is easier to collapse all of the products down into a single offering. It is designed to be a single point to contain all of the business logic necessary to create an Offering based on upstream data. It is somewhat similar to CandlepinPoolCapacityMapper in several ways.

Reply:
Sounds good! Will do.